src/shared/toolkitError.ts

```diff
@@ -0,0 +1,79 @@
+/*!
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as telemetry from './telemetry/telemetry'
+import { TimeoutError } from './utilities/timeoutUtils'
+
+interface ErrorMetadata {
+    // TODO: when `cause` is natively supported this can be removed
+    /**
+     * A reason for the error. This can be a string or another error.
+     */
+    readonly cause?: string | Error | ToolkitError
+
+    /**
+     * Detailed information about the error. This may be added to logs.
+     */
+    readonly detail?: string
+
+    /**
+     * Flag to determine if the error was from a user-initiated cancellation.
+     */
+    readonly cancelled?: boolean
+
+    readonly metricName?: string
+
+    /**
+     * Metric metadata associated with the error.
+     */
+    readonly metric?: {
+        /**
+         * The telemetry metric ID associated with this error.
+         *
+         * For example, if S3's `downloadFile` fails then we should use `s3_downloadObject` here.
+         */
+        readonly name: string
+
+        /**
+         * This should be set to either 'Failed' or 'Cancelled' depending on the control flow.
+         */
+        readonly result: Exclude<telemetry.Result, 'Succeeded'>
+    }
+}
+
+/**
+ * Error class for user-facing messages along with extra metadata.
+ */
+export class ToolkitError extends Error implements ErrorMetadata {
```

Comment:
Just an idea to get more general information out of errors while also providing a common mechanism for delivering user-facing error messages. Telemetry + error notification + log statement is an extremely common pattern. This error encapsulates all three ideas.